src/lib/operators/table_scan.hpp

```diff
@@ -25,7 +28,165 @@ class TableScan : public AbstractOperator {
   const AllTypeVariant& search_value() const;
 
  protected:
+  ColumnID _column_id;
+  ScanType _scan_type;
+  AllTypeVariant _search_value;
+
   std::shared_ptr<const Table> _on_execute() override;
+
+ private:
+  std::shared_ptr<std::vector<RowID>> _collect_matched_rows(std::shared_ptr<const Table>& input_table) const;
+  void _collect_matched_rows_for_segment(std::shared_ptr<BaseSegment>& segment, std::string& column_type,
+                                         ChunkID chunk_id, std::shared_ptr<std::vector<RowID>>& matched_row_ids) const;
+
+  std::shared_ptr<const Table> _determine_real_table(std::shared_ptr<const Table>& input_table) const;
+  std::shared_ptr<Table> _create_table_for_matched_rows(std::shared_ptr<std::vector<RowID>>& matched_row_ids,
+                                                        std::shared_ptr<const Table>& real_input_table) const;
+
+  template <typename T>
+  bool _evaluate_scan_predicate(const T& search_value, const T& value) const {
+    switch (_scan_type) {
```

Comment:
The switch statement gets evaluated for every loop iteration. Initially, we wanted to avoid that.